active/0015-unified-authentication.md

```diff
@@ -0,0 +1,169 @@
+# Unified Authentication in EMQ X 5.0
+
+## Change log
+
+*             @tigercl Initial draft
+* 2021-10-04: @zmstone Moved the doc from internal share
+
+## Abstract
+
+This proposal introduces a new design for EMQ X 5.0 authentication,
+which aims to provide: for EMQ X users, a better user experience with more configurable interfaces,
+and for EMQ X developers, a better development framework without repeating themselves.
+
+## Motivation
+
+EMQ X authentication is implemented by the hook-callback for hook-point `client.authenticate`.
+Up to v4.3, EMQ X supported 8 different authentication plugins, namely:
+
+```
+emqx_auth_http
+emqx_auth_jwt
+emqx_auth_ldap
+emqx_auth_mnesia
+emqx_auth_mongo
+emqx_auth_mysql
+emqx_auth_pgsql
+emqx_auth_redis
+```
+
+Some of the pain points in the old implementation
+
+1. The authentication plugins are implemented more or less the same,
+   and works more or less the same too. There is a lack of abstraction for the common parts,
+   causing developers to repeat themselves when adding new features or fixing issues.
+1. There is a lack of a nice re-configure interface, the only way to configure a plugin
+   is to update the config file, and reload (stop, start) the plugin.
+1. If there are more than one auth plugin enabled, there is no deterministic order for
+   how the different backends are checked.
+1. Enabled authentication plugins are collectively considered one global instance,
+   there is a lack of granularity for scoped control levels. e.g. per-zone, or per-listener.
+
+To address the pain-points in 5.0, we propose below enhancements.
+
+## Design
+
+### One app for all
+
+One `emqx_authn` app to unify the management of all different backends (except for ldap being postponed for now).
+
+### The same hook-point
+
+In this design, there is no intention to change the how EMQ X hooks work, the new app `emqx_authn` will
+continue to make use of the `client.authenticate` hook point,
+only to dispatch auth requests to the underlying backends inside one single hook call.
+
+### Composable authn "chain"
+
+We should allow users to compose (configure) a "chain" of backends with a defined order in which the
+checks are performed one after another. Each check against the backend in the chain may yield 3 different
+results for one-request authentication:
+
+- `ignore` is to indicate there is no auth information found hence should
+   continue validate the client against the rest of the backends in the chain.
+- `{ok, Info}` as a login accepted, hence to terminate the auth calls from here,
+   where `Info` may contain additional information such as to indicate if the user is a super-user.
+- `{error, Reason}` to indicate that client's login should be denied.
```

Comment:
nit, maybe it is due to backward backward compability,
I prefer more understandable ret values such like
- accept
- {reject, Reason}
- contitune 
- {error, Reason} %% when backend has other issues.

And also a default policy should be set on the chain such like `reject` or `accept`

Reply:
there is never going to be a default `accept`, if all auth backends `ignore`, the result is `not_authorized`.

I agree `accept` `reject` and `continue` are better indicators, @savonarola @tigercl please consider a refactoring.